### PR TITLE
Feat tensorboard

### PR DESCRIPTION
**Pull request by bernardo-marques**
Implemented tensorboard logging.
Should be compatible with previous configurations (i.e. no logging if the logging variable is missing)

Key changes:

1. Added tensorboardLog to the parameters of trainConfig
2. Added create_tensorboard_loggers() to trainSession
3. Added tensorboard_loggers=None as an input for do_training() (training.py)
4. Added a self.tensorboard_logger variable to the AccuracyOfEpochMonitorSegmentation class
5. Implemented AccuracyOfEpochMonitorSegmentation.log_to_tensorboard() to log dictionaries with (metric_name: metric_value) pairs
6. Changed AccuracyOfEpochMonitorSegmentation.reportAccuracyForLastSubepoch() to call AccuracyOfEpochMonitorSegmentation.log_to_tensorboard()
7. Changed inferenceWholeVolumes() (testing.py) to return a metrics dictionary to have DICE{1,2,3} be logged
8. Created TensorboardLogger class to remove tensorflow imports from accuracyMonitor.py

### CODE REVIEW

File: examples/configFiles/deepMedic/train/trainConfig.cfg [hunk 1]
```diff
@@ -61,7 +63,7 @@ typeOfSamplingForTraining = 3
 #  How many epochs to train for. Default: 35
 numberOfEpochs = 35
 #  How many subepochs comprise an epoch. Every subepoch I get Accuracy reported. Default: 20
-numberOfSubepochs = 20
+numberOfSubepochs = 1
```
Comment (Kamnitsask) [inline examples/configFiles/deepMedic/train/trainConfig.cfg]: This changes the default trainConfig arbitrarily. Nothing to do with pr change. Avoid such changes, they can go unnoticed.
```suggestion
numberOfSubepochs = 20
```

File: examples/configFiles/deepMedic/train/trainConfig.cfg [hunk 2]
```diff
@@ -193,7 +195,7 @@ typeOfSamplingForVal = 1
 
 #  +++++ Validation fully segmenting whole volumes +++++
 #  [Optional] How often (epochs) to perform validation by fully inferring validation volumes. Time consuming. Default: 1
-numberOfEpochsBetweenFullInferenceOnValImages = 5
+numberOfEpochsBetweenFullInferenceOnValImages = 1
```
Comment (Kamnitsask) [inline examples/configFiles/deepMedic/train/trainConfig.cfg]: This changes the default trainConfig arbitrarily. Nothing to do with pr change. Avoid such changes, they can go unnoticed.
```suggestion
numberOfEpochsBetweenFullInferenceOnValImages = 5
```

File: deepmedic/neuralnet/trainer.py [hunk 3]
```diff
@@ -254,7 +254,7 @@ def _get_mom_from_schedule(self):
             x2 = final_it_for_sch - first_it_for_sch
             x = tf.maximum( tf.constant(0, dtype="float32"), curr_it - first_it_for_sch )
             x = tf.minimum( x, x2 )
-            y_min = self._init_lr_tfv
+            y_min = self._init_mom_tfv
```
Comment (Kamnitsask) [inline deepmedic/neuralnet/trainer.py]: Changed  hotfixed in master, unrelated to current PR. Should have been pulled from master. There will be conflict.
Easy to solve but just saying. Better do changes only associated to PR and pull the rest.

File: deepmedic/dataManagement/io.py [hunk 4]
```diff
@@ -27,7 +27,7 @@ def loadVolume(filepath):
         img = np.expand_dims(img, axis=2)
     elif len(img.shape) > 3 :
         # 4D volumes could have been given. Often 3Ds are stored as 4Ds with 4th dim == 1.
-        assert img.shape[3] > 1
+        assert img.shape[3] == 1
```
Comment (Kamnitsask) [inline deepmedic/dataManagement/io.py]: This has been hotfixed on master. Should have been pulled. I d suggest each PR sticks with its own changes and pulls the rest from other branches. (eg tensorboard-branch may take a while to merge in master. This hotfix is easy to do in a more fast-to-merge branch, eg straight on dev or master). This will conflict during merge, but easy to resolve. I am mostly saying it for the future, as a practice.

File: examples/configFiles/deepMedic/train/trainConfig.cfg [hunk 5]
```diff
@@ -11,6 +11,8 @@ folderForOutput = "../../../output/"
 #  [Optional] Path to a saved model, to load parameters from at beginning of the session. If one is also specified from command line, the latter will be used.
 #cnnModelFilePath = "../../../output/models/placeholder"
 
+#  [Optional] Log to Tensorboard, set to 0 or remove to suppress logging to tensorboard
```
Comment (Kamnitsask) [inline examples/configFiles/deepMedic/train/trainConfig.cfg]: Rephrase more accurately.
```suggestion
#  [Optional] Log performance metrics for Tensorboard at end of subepochs. Give value 0 or remove this line to disable.
```

File: examples/configFiles/deepMedic/train/trainConfig.cfg [hunk 6]
```diff
@@ -11,6 +11,8 @@ folderForOutput = "../../../output/"
...
+tensorboardLog = 1
```
Comment (Kamnitsask) [inline examples/configFiles/deepMedic/train/trainConfig.cfg]: First, I think since this can only take two values, True or False, we may as well make them just give those. Instead of arbitrary numbers. Small difference, but checking.

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 7]
```diff
@@ -7,346 +7,455 @@
...
-        abs_path_to_saved = getAbsPathEvenIfRelativeIsGiven( cfg[cfg.SAVED_MODEL], abs_path_to_cfg ) if cfg[cfg.SAVED_MODEL] is not None else None # Load pretrained model.
-        self.savedModelFilepath = check_and_adjust_path_to_ckpt( self.log, abs_path_to_saved) if abs_path_to_saved is not None else None
-        
-        #====================TRAINING==========================
+        abs_path_to_saved = getAbsPathEvenIfRelativeIsGiven(cfg[cfg.SAVED_MODEL], abs_path_to_cfg) if cfg[
+                                                                                                          cfg.SAVED_MODEL] is not None else None  # Load pretrained model.
+        self.savedModelFilepath = check_and_adjust_path_to_ckpt(self.log,
+                                                                abs_path_to_saved) if abs_path_to_saved is not None else None
+
+        self.tensorboardLog = bool(cfg[cfg.TENSORBOARD_LOG])
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: Very concise, but allows values outside {0,1}. E.g. bool(99) is True. I would either make them give either True/False/None, or here check for arbitrary values and raise errors. Probably True/False is clearer.

Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]:
```suggestion
        self.tensorboardLog = bool(cfg[cfg.TENSORBOARD_LOG]) # If not given, bool(None) returns False (serves as default)
```

Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: Value of self.tensorboardLog is not logged at print_params(...).

In general, when we add a new parameter to the train/test/model config files, we also make sure to print it, so that its value is logged. So that in future, we can go back to the log files and make sure that the session was performed with the values we wanted it to.

See Line 458+

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 8]
```diff
@@ -357,42 +466,53 @@ def print_params(self) :
         logPrint("~~Output~~")
         logPrint("Main output folder = " + str(self.mainOutputAbsFolder))
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: Print/log the new variable:

```suggestion
        logPrint("Main output folder = " + str(self.mainOutputAbsFolder))
        logPrint("Log performance metrics for tensorboard = " + str(self.tensorboardLog))
```

File: examples/configFiles/deepMedic/train/trainConfig.cfg (hunk 5 above)
Comment (Kamnitsask) [inline examples/configFiles/deepMedic/train/trainConfig.cfg]: Also, I would either put this right after the "output folder", or at the far end of the file. Not sure.

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 9]
```diff
@@ -7,346 +7,455 @@
...
-                               'auto': { 'min_incr_of_val_acc_considered': cfg[cfg.AUTO_MIN_INCR_VAL_ACC] if cfg[cfg.AUTO_MIN_INCR_VAL_ACC] is not None else 0.0,
-                                         'epochs_wait_before_decr': cfg[cfg.NUM_EPOCHS_WAIT] if cfg[cfg.NUM_EPOCHS_WAIT] is not None else 5,
-                                         'div_lr_by': cfg[cfg.DIV_LR_BY] if cfg[cfg.DIV_LR_BY] is not None else 2.0 },
-                               'poly': { 'epochs_wait_before_decr': cfg[cfg.NUM_EPOCHS_WAIT] if cfg[cfg.NUM_EPOCHS_WAIT] is not None else self.numberOfEpochs/3,
-                                         'final_ep_for_sch': self.numberOfEpochs },
-                               'expon': { 'epochs_wait_before_decr': cfg[cfg.NUM_EPOCHS_WAIT] if cfg[cfg.NUM_EPOCHS_WAIT] is not None else self.numberOfEpochs/3,
+                                'predef': {'epochs': cfg[cfg.PREDEF_SCH],
+                                           'div_lr_by': cfg[cfg.DIV_LR_BY] if cfg[cfg.DIV_LR_BY] is not None else 2.0},
+                                'auto': {'min_incr_of_val_acc_considered': cfg[cfg.AUTO_MIN_INCR_VAL_ACC] if cfg[
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: These changes from PEP breaking "if cfg[cfg...]" are quite ugly. I would undo these here to be honest.

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 10]
```diff
@@ -7,346 +7,455 @@
...
         self.augm_sample_prms_tr = {'hist_dist': None, 'reflect': None, 'rotate90': None}
         if cfg[cfg.AUGM_SAMPLE_PRMS_TR] is not None:
             for key in cfg[cfg.AUGM_SAMPLE_PRMS_TR]:
-                self.augm_sample_prms_tr[key] = cfg[cfg.AUGM_SAMPLE_PRMS_TR][key] # For exact form of parameters, see ./deepmedic/dataManagement/augmentation.py
-        
-        #===================VALIDATION========================
-        self.val_on_samples_during_train = cfg[cfg.PERFORM_VAL_SAMPLES] if cfg[cfg.PERFORM_VAL_SAMPLES] is not None else False
-        if self.lr_sched_params['type'] == 'auto' and not self.val_on_samples_during_train :
+                self.augm_sample_prms_tr[key] = cfg[cfg.AUGM_SAMPLE_PRMS_TR][
+                    key]  # For exact form of parameters, see ./deepmedic/dataManagement/augmentation.py
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: weird line breaks. I will undo some of them at a future commit

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 11]
```diff
@@ -7,346 +7,455 @@
...
+            self.eAdam = cfg[cfg.EPS_ADAM] if cfg[cfg.EPS_ADAM] is not None else 10 ** (-8)
+            self.rhoRms = "placeholder";
+            self.eRms = "placeholder";
+        elif self.optimizerSgd0Adam1Rms2 == 2:
+            self.b1Adam = "placeholder";
+            self.b2Adam = "placeholder";
+            self.eAdam = "placeholder";
+            self.rhoRms = cfg[cfg.RHO_RMS] if cfg[cfg.RHO_RMS] is not None else 0.9  # default in paper and seems good
+            self.eRms = cfg[cfg.EPS_RMS] if cfg[cfg.EPS_RMS] is not None else 10 ** (
+                -4)  # 1e-6 was the default in the paper, but blew up the gradients in first try. Never tried 1e-5 yet.
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: weitrd linebreak. Just move comment I guess.

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 12]
```diff
@@ -7,346 +7,455 @@
...
+            else:  # Only one case. Just give the output prediction folder, the io.py will save output accordingly.
+                self.filepathsToSavePredictionsForEachPatientVal.append(absPathToFolderForPredictionsFromSession)
+                self.filepathsToSaveFeaturesForEachPatientVal.append(absPathToFolderForPredictionsFromSession)
+
     def get_path_to_load_model_from(self):
         return self.savedModelFilepath
-    
-    def print_params(self) :
+
+    def get_tensorboard_bool(self):
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: # Called only by ./frontEnd/trainSession.py. Can be avoided? I will have a look

File: deepmedic/frontEnd/configParsing/trainSessionParams.py [hunk 13]
```diff
@@ -7,346 +7,455 @@
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/configParsing/trainSessionParams.py]: Dont run + commit PEP in the same commit as actual changes. Makes it impossible to find them during code review. When running PEP, just make it a completely separate commit.

File: deepmedic/routines/testing.py [hunk 14]
```diff
@@ -404,7 +404,13 @@ def inferenceWholeVolumes(  sessionTf,
     log.print3("###########################################################################################################")
     log.print3("############################# Finished full Segmentation of " + str(validation_or_testing_str) + " subjects ##########################")
     log.print3("###########################################################################################################")
-    return 0
+
+    metrics_dict_list = []
+    for i in range(len(meanDiceCoeffs1)):
```
Comment (Kamnitsask) [inline deepmedic/routines/testing.py]: If ground truth labels are not given (e.g. for testing) then Dice Coffs are undefined. Just inference is done and results are saved.

In that case, error occurs. E.g. try:
python ./deepMedicRun -model examples/configFiles/tinyCnn/model/modelConfig.cfg -test examples/configFiles/tinyCnn/test/testConfig.cfg -load some-pretrained-model

I think this can be avoided if these are moved 5 lines further up, at the end of the block:
if listOfFilepathsToGtLabelsOfEachPatient is not None and total_number_of_images>0 :

File: deepmedic/frontEnd/trainSession.py [hunk 15]
```diff
@@ -19,148 +19,186 @@
...
             self._log.print3("=======================================================")
             self._log.print3("============== Training the CNN model =================")
             self._log.print3("=======================================================\n")
-            
-            res_code = do_training( *( [sessionTf, saver_all, cnn3d, trainer] + self._params.get_args_for_train_routine() ) )
-            
-            
+
+            do_training(*([sessionTf, saver_all, cnn3d, trainer] + self._params.get_args_for_train_routine()),
+                        tensorboard_loggers=tensorboard_loggers)
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/trainSession.py]: I would move it within the [...], which holds all the tensorflow parameters:

[sessionTf, saver_all, cnn3d, trainer, tensorboard_loggers]

File: deepmedic/frontEnd/trainSession.py [hunk 16]
```diff
@@ -19,148 +19,186 @@
...
-    def _print_vars_in_collection(self, collection, coll_name="no_name"):
+         self._out_folder_fms] = makeFoldersNeededForTrainingSession(self._main_out_folder_abs, self._sess_name)
+
+    def create_tensorboard_loggers(self, logger_types, tf_graph, create_log=False):
+        tensorboard_loggers = {}
+
+        self._log.print3("----------- Creating Tensorboard Loggers -----------")
+
+        for logger_type in logger_types:
+            if create_log:
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/trainSession.py]: repeated if-create_log-else with later. Can be merged (along with the prints).

File: deepmedic/frontEnd/trainSession.py [hunk 17]
```diff
@@ -19,148 +19,186 @@
...
+                tensorboard_loggers[logger_type] = TensorboardLogger(os.path.join(self._main_out_folder_abs,
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/trainSession.py]: Create the foldeer in:
makeFoldersNeededForTrainingSession (L46 above)

File: deepmedic/logging/tensorboard_logger.py [hunk 18]
```diff
@@ -0,0 +1,13 @@
+from __future__ import absolute_import, print_function, division
+
+import tensorflow as tf
+
+
+class TensorboardLogger(object):
+
+    def __init__(self, log_path, tf_graph):
+        self.logger = tf.compat.v1.summary.FileWriter(log_path, tf_graph)
```
Comment (Kamnitsask) [inline deepmedic/logging/tensorboard_logger.py]: This line threw error with Tensorflow 1.12.0 (no such module).

Why not:
```suggestion
        self.logger = tf.summary.FileWriter(log_path, tf_graph)
```
??

This one works fine for me. Not sure if something changes in later TF versions.

File: deepmedic/frontEnd/trainSession.py [hunk 19]
```diff
@@ -19,148 +19,186 @@
...
+            # Because trained has piecewise_constant that is only on cpu, and so is saver.
             with tf.variable_scope("trainer"):
                 self._log.print3("=========== Building Trainer ===========\n")
-                trainer = Trainer( *( self._params.get_args_for_trainer() + [cnn3d] ) )
-                trainer.create_optimizer( *self._params.get_args_for_optimizer() ) # Trainer and net connect here.
-                
+                trainer = Trainer(*(self._params.get_args_for_trainer() + [cnn3d]))
+                trainer.create_optimizer(*self._params.get_args_for_optimizer())  # Trainer and net connect here.
+
+                tensorboard_loggers = self.create_tensorboard_loggers(['train', 'val'],
```
Comment (Kamnitsask) [inline deepmedic/frontEnd/trainSession.py]: Does not need an initializer, nor a saver.

I would definitely not put it in a "net" or "trainer" scope. Lets keep those clean for very specific variables.

It can just come out of scope (one tab in the left), or if we want to be more explicit, lets make an "others" scope:
```suggestion
            with tf.variable_scope("others"):
                tensorboard_loggers = self.create_tensorboard_loggers(['train', 'val'],
```

File: deepmedic/frontEnd/configParsing/trainSessionParams.py (hunk 12 above)
Comment (Kamnitsask) [reply deepmedic/frontEnd/configParsing/trainSessionParams.py]: I cannot be avoided.

File: deepmedic/logging/tensorboard_logger.py (hunk 18 above)
Comment (Kamnitsask) [inline deepmedic/logging/tensorboard_logger.py]: I would add a note (for us) here that says that this does NOT add a variable to the graph, and thus does not need initialization.

(can be checked by:
```
with tf.variable_scope("others"):
                tensorboard_loggers = self.create_tensorboard_loggers(['train', 'val'], # Main Difference 2.
                                                                      graphTf,
                                                                      create_log=self._params.get_tensorboard_bool())

collection_vars_others = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope="others")
print(collection_vars_others)    # Returns empty list []
```

File: deepmedic/routines/training.py [hunk 20]
```diff
@@ -24,382 +24,437 @@
...
-            
-            acc_monitor_for_ep_train = AccuracyOfEpochMonitorSegmentation(log, 0, model_num_epochs_trained, cnn3d.num_classes, num_subepochs)
+
+            acc_monitor_for_ep_train = \
+                AccuracyOfEpochMonitorSegmentation(log, 0,
+                                                   model_num_epochs_trained,
+                                                   cnn3d.num_classes,
+                                                   num_subepochs,
+                                                   tensorboard_logger=tensorboard_loggers['train']
+                                                   if tensorboard_loggers is not None else None)
```
Comment (Kamnitsask) [inline deepmedic/routines/training.py]: Kamnitsask 16 seconds ago Member

This if here inside the arguments, in its own line, reads "very" weird.

If we rename "tensorboard_loggers" to "tb_loggers" it can go to same line probably.

File: deepmedic/routines/training.py [hunk 21]
```diff
@@ -24,382 +24,437 @@
...
+
             acc_monitor_for_ep_val = None if not val_on_samples_during_train else \
-                                            AccuracyOfEpochMonitorSegmentation(log, 1, model_num_epochs_trained, cnn3d.num_classes, num_subepochs )
-                                            
+                AccuracyOfEpochMonitorSegmentation(log, 1,
+                                                   model_num_epochs_trained,
+                                                   cnn3d.num_classes,
+                                                   num_subepochs,
+                                                   tensorboard_logger=tensorboard_loggers['val']
+                                                   if tensorboard_loggers is not None else None)
```
Comment (Kamnitsask) [inline deepmedic/routines/training.py]: ditto

File: deepmedic/routines/training.py [hunk 22]
```diff
@@ -24,382 +24,437 @@
...
             epoch = model_num_epochs_trained
-            
-            acc_monitor_for_ep_train = AccuracyOfEpochMonitorSegmentation(log, 0, model_num_epochs_trained, cnn3d.num_classes, num_subepochs)
+
+            acc_monitor_for_ep_train = \
+                AccuracyOfEpochMonitorSegmentation(log, 0,
+                                                   model_num_epochs_trained,
+                                                   cnn3d.num_classes,
+                                                   num_subepochs,
+                                                   tensorboard_logger=tensorboard_loggers['train']
```
Comment (Kamnitsask) [inline deepmedic/routines/training.py]: probably this argument will look better right after the "log" argument

File: deepmedic/routines/training.py [hunk 23]
```diff
@@ -24,382 +24,437 @@
...
+                    saveIndividualFmImagesForVisualisation=saveIndividualFmImagesForVisualisation,
+                    saveMultidimensionalImageWithAllFms=saveMultidimensionalImageWithAllFms,
+                    indicesOfFmsToVisualisePerPathwayTypeAndPerLayer=indicesOfFmsToVisualisePerPathwayTypeAndPerLayer,
+                    namesForSavingFms=namesForSavingFms
+                )
+
+                acc_monitor_for_ep_val.reportDSCWholeSegmentation(metrics_dict_list)
+
+            del acc_monitor_for_ep_train
+            del acc_monitor_for_ep_val
```
Comment (Kamnitsask) [inline deepmedic/routines/training.py]: I think we can safely delete these. The garbage collector will do the job anyway.

File: deepmedic/logging/accuracyMonitor.py [hunk 24]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+            class_string = classString
+            self.log_to_tensorboard(metrics_dict, class_string, step_num)
+
+        self.log.print3('Logged metrics: ' + str(list(metrics_dict.keys())))
+        self.log.print3('======================================================')
+
+    def reportDSCWholeSegmentation(self, metrics_dict_list):
+        if self.tensorboard_logger is not None:
+            self.log.print3('=============== LOGGING TO TENSORBOARD ===============')
+            self.log.print3('Logging whole brain segmentation metrics')
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: It is not brain segmentation specific

```suggestion
            self.log.print3('Logging validation metrics from segmentation of whole validation volumes.')
```

File: deepmedic/routines/training.py [hunk 25]
```diff
@@ -24,382 +24,437 @@
...
+                    pad_input_imgs=pad_input_imgs,
+
+                    # --------For FM visualisation---------
+                    saveIndividualFmImagesForVisualisation=saveIndividualFmImagesForVisualisation,
+                    saveMultidimensionalImageWithAllFms=saveMultidimensionalImageWithAllFms,
+                    indicesOfFmsToVisualisePerPathwayTypeAndPerLayer=indicesOfFmsToVisualisePerPathwayTypeAndPerLayer,
+                    namesForSavingFms=namesForSavingFms
+                )
+
+                acc_monitor_for_ep_val.reportDSCWholeSegmentation(metrics_dict_list)
```
Comment (Kamnitsask) [inline deepmedic/routines/training.py]: perhaps rename this to the below? Its functionality currently looks more like that.
log_acc_from_whole_volume_segm_to_tensorboard

(while the "reportSomethingSomething" functions write to the txt log.

File: deepmedic/logging/accuracyMonitor.py [hunk 26]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+    def log_to_tensorboard(self, metrics_dict, class_string, step_num):
+        if self.tensorboard_logger is not None:
+            for metric, value in metrics_dict.items():
+                if value == self.NA_PATTERN:
+                    value = np.nan
+                self.tensorboard_logger.add_summary(value, metric + '_' + class_string, step_num)
+
+    def log_acc_subep_to_tensorboard(self):
+
+        trainOrValString = "TRAINING" if self.training0orValidation1 == 0 else "VALIDATION"
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: can become a self.trainOrValString , as it's used in 2 functions now

File: deepmedic/logging/accuracyMonitor.py [hunk 27]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+
+        self.log.print3('=============== LOGGING TO TENSORBOARD ===============')
+        self.log.print3('Logging ' + trainOrValString + ' metrics')
+        self.log.print3('Epoch: ' + str(self.epoch) +
+                        ' | Subepoch ' + str(currSubep) + '/' + str(self.numberOfSubepochsPerEpoch - 1))
+        step_num = currSubep + (self.epoch * (self.numberOfSubepochsPerEpoch))
+        self.log.print3('Step number: ' + str(step_num))
+
+        # check if user included tensorboard logging in the config
+        if self.tensorboard_logger is None:
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: I 'd put this right under the "logging to tensorboard", to avoid printing epoch/subepoch etc if it's deactivated. Less logging. (eg "step number" does not make sense).

File: deepmedic/frontEnd/configParsing/trainSessionParams.py (hunk 7 above)
Comment (bernardo-marques) [reply deepmedic/frontEnd/configParsing/trainSessionParams.py]: I remember that at the time I had issues with boolean inputs, and that's why I used the {1,0}. I agree True/False is better

File: deepmedic/frontEnd/configParsing/trainSessionParams.py (hunk 9 above)
Comment (bernardo-marques) [reply deepmedic/frontEnd/configParsing/trainSessionParams.py]: I think this is part of the Pycharm automatic cleanup that I missed. I agree there's a better way of doing it, but we should still keep the line width to a limit for ease of reading

File: deepmedic/logging/accuracyMonitor.py [hunk 28]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+
+        trainOrValString = "TRAINING" if self.training0orValidation1 == 0 else "VALIDATION"
+        currSubep = self.numberOfSubepochsForWhichUpdated - 1
+
+        self.log.print3('=============== LOGGING TO TENSORBOARD ===============')
+        self.log.print3('Logging ' + trainOrValString + ' metrics')
+        self.log.print3('Epoch: ' + str(self.epoch) +
+                        ' | Subepoch ' + str(currSubep) + '/' + str(self.numberOfSubepochsPerEpoch - 1))
+        step_num = currSubep + (self.epoch * (self.numberOfSubepochsPerEpoch))
+        self.log.print3('Step number: ' + str(step_num))
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: I d make "step number" more descriptive. Something like: "Logging for step (x-axis, subepochs trained): #
I could even rename this "step_num" to x_value

File: deepmedic/logging/accuracyMonitor.py [hunk 29]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+             meanDiceOfSubep] = self.listPerSubepPerClassMeanAccSensSpecDsc[currSubep][class_i] if class_i != 0 else \
+                self.listPerSubepForegrMeanAccSensSpecDsc[currSubep]  # If class-0, report foreground.
+
+            # create metrics dictionary
+            metrics_dict = {'mean_acc': meanAccClassOfSubep,
+                            'mean_sens': meanAccOnPosOfSubep,
+                            'mean_prec': meanPrecOfSubep,
+                            'mean_spec': meanAccOnNegOfSubep,
+                            'mean_dice': meanDiceOfSubep}
+            class_string = classString
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: get rid of this, by changing L201 and L131

File: deepmedic/logging/accuracyMonitor.py [hunk 30]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+
+    def log_acc_subep_to_tensorboard(self):
+
+        trainOrValString = "TRAINING" if self.training0orValidation1 == 0 else "VALIDATION"
+        currSubep = self.numberOfSubepochsForWhichUpdated - 1
+
+        self.log.print3('=============== LOGGING TO TENSORBOARD ===============')
+        self.log.print3('Logging ' + trainOrValString + ' metrics')
+        self.log.print3('Epoch: ' + str(self.epoch) +
+                        ' | Subepoch ' + str(currSubep) + '/' + str(self.numberOfSubepochsPerEpoch - 1))
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: dont subtract from self.numberOfSubepochsPerEpoc - 1, cause then if you specify 20 epochs, it shows x/19.
x=curr_subepoch currently starts count from 0. While self.numberOf is 1...20.
I understand this is annoying.
But we should somehow fix how we print curr_subepoch. Not (wrongly) change the self.numberOfSubepochs.

File: deepmedic/logging/accuracyMonitor.py [hunk 31]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+                            strFl4fNA(meanAccClassOfSubep, self.NA_PATTERN) +
+                            "\t=> (TruePos+TrueNeg)/All-Predicted-Voxels = " +
+                            str(numOfTpInSubep+numOfTnInSubep) + "/" +
+                            str(numOfRpInSubep+numOfRnInSubep))
+            self.log.print3(logStrClass+"\t mean sensitivity:\t" + strFl4fNA(meanAccOnPosOfSubep, self.NA_PATTERN)+"\t=> TruePos/RealPos = "+str(numOfTpInSubep)+"/"+str(numOfRpInSubep))
+            self.log.print3(logStrClass+"\t mean precision:\t" + strFl4fNA(meanPrecOfSubep, self.NA_PATTERN)+"\t=> TruePos/(TruePos+FalsePos) = "+str(numOfTpInSubep)+"/"+str(numOfTpInSubep+numOfFpInSubep))
+            self.log.print3(logStrClass+"\t mean specificity:\t" + strFl4fNA(meanAccOnNegOfSubep, self.NA_PATTERN)+"\t=> TrueNeg/RealNeg = "+str(numOfTnInSubep)+"/"+str(numOfRnInSubep))
+            self.log.print3(logStrClass+"\t mean Dice:       \t" + strFl4fNA(meanDiceOfSubep, self.NA_PATTERN))
+
+    def log_to_tensorboard(self, metrics_dict, class_string, step_num):
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: log_to_tensorboard_ys_for_x(...)

File: deepmedic/logging/accuracyMonitor.py [hunk 32]
```diff
@@ -146,12 +148,91 @@ def reportAccuracyForLastSubepoch(self) :
...
+                            'mean_prec': meanPrecOfSubep,
+                            'mean_spec': meanAccOnNegOfSubep,
+                            'mean_dice': meanDiceOfSubep}
+            class_string = classString
+            self.log_to_tensorboard(metrics_dict, class_string, step_num)
+
+        self.log.print3('Logged metrics: ' + str(list(metrics_dict.keys())))
+        self.log.print3('======================================================')
+
+    def reportDSCWholeSegmentation(self, metrics_dict_list):
```
Comment (Kamnitsask) [inline deepmedic/logging/accuracyMonitor.py]: When a function needs some non crystal clear data-structure, describe it in a comment (eg the metrics_dict_list here)

Comment (Kamnitsask) [reply deepmedic/logging/accuracyMonitor.py]: metrics_dict_list: list holds one element per class, [ elem-class0, elem-class1, .... ]
Each element is itself a dictionary of metrics for the class.
E.g. elem-class0 = {'mean_dice1': value, 'mean_dice2': value, mean_dice3': value}

File: deepmedic/logging/loggers.py [hunk 33]
```diff
@@ -9,18 +9,104 @@
...
+
+            self.print3(string)
+
+        if block:
+            self.print_pattern_line(pattern, line_width)
         
-    def __init__(self, filenameAndPathOfLoggerTxt="logs/defaultLogFile.txt") :
-        self.loggerFileName = filenameAndPathOfLoggerTxt
+    def __init__(self, log_path="logs/defaultLogFile.txt"):
+        self.loggerFileName = log_path
```
Comment (Kamnitsask) [inline deepmedic/logging/loggers.py]: Only last camelCase variable. We may as well change it to self.log_path

File: deepmedic/routines/testing.py (hunk 14 above)
Comment (bernardo-marques) [reply deepmedic/routines/testing.py]: fixed as suggested

File: deepmedic/frontEnd/trainSession.py (hunk 15 above)
Comment (bernardo-marques) [reply deepmedic/frontEnd/trainSession.py]: I think we should make a class that holds all the parameters and parse that, but that's a bigger structural task. I'd leave it for the

File: deepmedic/frontEnd/trainSession.py (hunk 17 above)
Comment (bernardo-marques) [reply deepmedic/frontEnd/trainSession.py]: Let's discuss how to best approach this.

File: deepmedic/logging/accuracyMonitor.py (hunk 31 above)
Comment (bernardo-marques) [reply deepmedic/logging/accuracyMonitor.py]: I think this is unnecessary. A tensorboard log is always associated with a step, and a log to tensorboard is always of metrics

File: deepmedic/logging/accuracyMonitor.py (hunk 30 above)
Comment (bernardo-marques) [reply deepmedic/logging/accuracyMonitor.py]: Currently if there's only one epoch iut will say 0/1, and never get to 1/1. If epoch 0 exists, then I think this is the best approach, otherwise it will not be matched with what actually happens

File: deepmedic/logging/loggers.py (hunk 33 above)
Comment (bernardo-marques) [reply deepmedic/logging/loggers.py]: change not made because it needs to be propagated in the rest of the code in a more structural renaming session

File: deepmedic/routines/training.py (hunk 22 above)
Comment (bernardo-marques) [reply deepmedic/routines/training.py]: that would make it  a very structural change, and create compatibility issues with branches of code that have changed that function.

File: deepmedic/routines/training.py (hunk 20 above)
Comment (bernardo-marques) [reply deepmedic/routines/training.py]: once the AccuracyOfEpochMonitorSegmentation fucntion has been renamed to something smaller this issue will be easily mitigated

File: deepmedic/routines/training.py (hunk 25 above)
Comment (bernardo-marques) [reply deepmedic/routines/training.py]: I just followed the current naming convention and left these changes to be done when restructuring the code. Mixed conventions look weird

File: deepmedic/frontEnd/trainSession.py (hunk 15 above)
Comment (Kamnitsask) [reply deepmedic/frontEnd/trainSession.py]: you mean because it has an optional arg
